core/services/vrf/listener_v2.go

```diff
@@ -0,0 +1,371 @@
+package vrf
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/gracefulpanic"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/vrf_coordinator_v2"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/bulletprooftxmanager"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	httypes "github.com/smartcontractkit/chainlink/core/services/headtracker/types"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/services/log"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/services/postgres"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/multierr"
+	"gopkg.in/guregu/null.v4"
+	"gorm.io/gorm"
+)
+
+const (
+	StaticFulfillExecuteGasCost = 5000 + // subID balance update
+		2100 + // cold subscription balance read
+		20000 + // first time oracle balance update, note first time will be 20k, but 5k subsequently
+		2*2100 - // cold read oracle address and oracle balance
+		15000 // request delete refund
+
+	// Buffer to ensure that the gas received after the call to the consumer contract is
+	// at least the amount they requested. Same argument as
+	// https://github.com/cholladay0816/chainlink/blob/08b6fd1b910b5e9b5d20f834a09204d159a56142/contracts/src/v0.6/VRFCoordinator.sol#L201
+	BufferForConsumerCallback = 6000
+
+	CallFulfillGasCost = 21000 + // Base tx cost
+		7317 // Static costs of argument encoding etc. note that it varies by +/- x*12 for every x bytes of non-zero data in the proof.
```

Comment:
We could actually make this exact if we want, at the expense of appending another variable to the proof containing the number of non-zero or zero bytes in the proof (I tried doing it onchain and it was surprisingly expensive ~100 gas per loop to do the comparison * ~608 loop iterations, could probably make it cheaper with assembly though).